File: src/testWithSpringBoot_2_4/java/org/openrewrite/java/spring/boot2/MergeBootstrapYamlWithApplicationYamlTest.java
/*
 * Copyright 2024 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.spring.boot2;

import org.junit.jupiter.api.Test;
import org.openrewrite.test.RecipeSpec;
import org.openrewrite.test.RewriteTest;

import static org.openrewrite.java.Assertions.srcMainResources;
import static org.openrewrite.test.SourceSpecs.other;
import static org.openrewrite.yaml.Assertions.yaml;

class MergeBootstrapYamlWithApplicationYamlTest implements RewriteTest {

    @Override
    public void defaults(RecipeSpec spec) {
        spec.recipe(new MergeBootstrapYamlWithApplicationYaml());
    }

    @Test
    void mergeBootstrap() {
        rewriteRun(
          srcMainResources(
            //language=yaml
            yaml(
              """
                spring.application.name: main
                """,
              """
                spring.application.name: main
                name: test
                """,
              spec -> spec.path("application.yaml")
            ),
            //language=yaml
            yaml(
              """
                name: test
                """,
              null,
              spec -> spec.path("bootstrap.yaml")
            )
          )
        );
    }

    @Test
    void mergeMultipleBootstrapDocuments() {
        rewriteRun(
          srcMainResources(
            //language=yaml
            yaml(
              """
                spring.application.name: main
                """,
              """
                spring.application.name: main
                name: test
                other.document: true
                """,
              spec -> spec.path("application.yaml")
            ),
            //language=yaml
            yaml(
              """
                name: test
                ---
                other:
                  document: true
                """,
              null,
              spec -> spec.path("bootstrap.yaml")
            )
          )
        );
    }

    @Test
    void createsApplicationYaml() {
        rewriteRun(
          srcMainResources(
            //language=yaml
            yaml(
              null,
              """
                spring.application.name: main
                name: test
                """,
              spec -> spec.path("application.yaml")
            ),
            //language=yaml
            yaml(
              """
                spring.application:
                  name: main
                name: test
                """,
              null,
              spec -> spec.path("bootstrap.yaml")
            )
          )
        );
    }

    @Test
    void doNotMergeExistingKeys() {
        rewriteRun(
          srcMainResources(
            //language=yaml
            yaml(
              """
                spring.application.name: main
                """,
              """
                spring.application.name: main
                """,
              spec -> spec.path("application.yaml")
            ),
            //language=yaml
            yaml(
              """
                spring.application:
                  name: override
                ---
                spring.application.name: override
                ---
                spring:
                  application:
                    name: override
                """,
              null,
              spec -> spec.path("bootstrap.yaml")
            )
          )
        );
    }

    @Test
    void doNotMergeProfileSpecificDocuments() {
        rewriteRun(
          srcMainResources(
            //language=yaml
            yaml(
              """
                spring:
                  application.name: main
                """,
              """
                spring.application.name: main
                name: test
                """,
              spec -> spec.path("application.yaml")
            ),
            //language=yaml
            yaml(
              """
                name: test
                ---
                spring.config.activate.on-profile: test
                name: profile-test
                other:
                  document: false
                """,
              null,
              spec -> spec.path("bootstrap.yaml")
            )
          )
        );
    }

    @Test
    void doNotMergeWhenNotValidApplicationYaml() {
        rewriteRun(
          srcMainResources(
            //language=yaml
            other("""
                spring:
                  application.name: main
                """,
              spec -> spec.path("application.yaml")),
            //language=yaml
            yaml(
              """
                name: test
                """,
              spec -> spec.path("bootstrap.yaml")
            )
          )
        );
    }

    @Test
    void doNotMergeWhenNotValidBootstrapYaml() {
        rewriteRun(
          srcMainResources(
            //language=yaml
            yaml("""
                spring:
                  application.name: main
                """,
              spec -> spec.path("application.yaml")),
            //language=yaml
            other(
              """
                name: test
                """,
              spec -> spec.path("bootstrap.yaml")
            )
          )
        );
    }
}

File: src/testWithSpringBoot_2_4/kotlin/org/openrewrite/java/spring/boot2/MergeBootstrapYamlWithApplicationYamlTest.kt
/*
 * Copyright 2021 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.spring.boot2

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.openrewrite.Recipe
import org.openrewrite.yaml.YamlParser
import org.openrewrite.yaml.YamlRecipeTest
import java.nio.file.Paths

class MergeBootstrapYamlWithApplicationYamlTest : YamlRecipeTest {
    override val recipe: Recipe
        get() = MergeBootstrapYamlWithApplicationYaml()

    @Test
    fun mergeBootstrap() {
        val applicationYaml = YamlParser().parse(
            """
                spring:
                  application.name: main
                ---
                spring:
                    config:
                        activate:
                            on-profile: test
                name: test
            """.trimIndent()
        ).map { it.withSourcePath(Paths.get("src/main/resources/application.yml")) }

        val bootstrapYaml = YamlParser().parse(
            """
                spring.application:
                  name: not the name
                ---
                spring:
                    config:
                        activate:
                            on-profile: test
                name: test
            """.trimIndent()
        ).map { it.withSourcePath(Paths.get("src/main/resources/bootstrap.yml")) }

        val results = recipe.run(bootstrapYaml + applicationYaml)
            .results.sortedBy { it.before!!.sourcePath.fileName.toString() }

        assertThat(results).hasSize(2)
        assertThat(results[1].after).isNull()
        assertThat(results[0].after!!.printAll()).isEqualTo(
            """
                spring.application.name: main
                ---
                spring.config.activate.on-profile: test
                name: test
            """.trimIndent()
        )
    }
}
